### ISSUE TITLE
Multiple outputs, same hidden structure

### ISSUE THREAD

**Issue by santi-pdp** [fchollet/keras#1320]
Hello people,

I have been working with a regression problem with LSTMs in which I predict a vector of output coefficients from an input sequence. What I want to do is to build multiple output layers on top of the same hidden structure, so that I will be training the different layers depending on the input sequence (inputs always have the same dimension). To sum up, when I backprop one output layer all other output layers will remain static. Is there any way to freeze the N-1 output layers and only fit a batch of data into the Nth layer, iterating the process for many batches and the N layers? Can I have any clue please?

Thank you

**Comment by sjebbara** [comment]
Hi,
as far as I know, you could build *N* different models that share the same layers up until your output layers, e.g. like:

**Note:** I did not run the code but I think something along these lines is possible with keras (at least with the Graph() model).
```python
# compose a "submodel" that contains all the layers that are shared across your N models
shared_model = Sequential()
# add all layers up until you last layers here, like so:
shared_model.add(LSTM(output_dim=100, input_dim=20, return_sequences=False))
shared_model.add(Dropout(0.5))
# other shared layers ...

# now combine the shared "submodel" with the individual output
# layers by creating separate "final" models:
model1 = Sequential()
model1.add(shared_model)
model1.add(Dense(10, activation="softmax"))

model2 = Sequential()
model2.add(shared_model)
model2.add(Dense(2, activation="linear"))
# ...

# build your N models
model1.compile("rmsprop", "categorical_crossentropy")
model2.compile("rmsprop", "mse")
# ...

# and train them alternatingly:
for X_batch, Y1_batch, Y2_batch, ... in your_data:
    model1.train_on_batch(X_batch, Y1_batch)
    model2.train_on_batch(X_batch, Y2_batch)
    # ...
```

Since the first layers are the same in all models, their weights are essentially shared across these models.
Is this what you were looking for?

**Comment by santi-pdp** [comment]
Well the second example is how I was thinking to proceed, but from the first you've put I have a doubt: when you specify multiple outputs in the graph and you feed the model with the dictionary of inputs/outputs, are the outputs trained separately? (thus freezing the other outputs) or are they all back-propagated at once, computing the weights for all outputs? I have used the Graph previously to have probabilistic outputs and regression outputs with different objective functions, but not to train different output layers separately. 
Thanks a lot!

**Comment by snakeztc** [comment]
Hi, @santi-pdp I have the exact the same question as you do. Which model (sequential or graph) did you finally pick? Do you have a good solution of freezing the the rest of output and only back-propagate 1 output? Thank you!

**Comment by mbalty** [comment]
Here is how I did it. Is this correct? 
`out = Dense(128, activation='relu', name='fcc')(out)
    out_reg = Dense(1, name='out_reg')(out)
    out_cls = Dense(n_cls, activation='softmax', name='out_cls')(out)
    
    model = Model(input=inp, output=[out_reg, out_cls])
    model_reg = Model(input=inp, output=out_reg)
    model_cls = Model(input=inp, output=out_cls)

    model.compile(loss=['mean_squared_error', 'categorical_crossentropy'], optimizer='adam', metrics=['accuracy'])
    model_reg.compile(loss='mean_squared_error', optimizer='adam')
    model_cls.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
    
    return model, model_reg, model_cls`

Output is the hidden layer, and then 2 outputs: _reg, _cls. Model is trained minimizing both

**Comment by malheirosrafa** [comment]
@mbalty What are the results you achieve? Can you share the full model (inputs, etc)?

**Comment by saxenarohit** [comment]
@mbalty was your implementation correct? Did you get the intended results?
